rust/pact_matching_ffi/ARCHITECTURE.md

```diff
@@ -0,0 +1,153 @@
+This design is based on the work of Michael F. Bryan in the
+[unofficial Rust FFI Guide][ffi_guide], which was based on the error handling
+in `libgit2`.
+
+In short: errors are collected from all FFI code, and stored in a
+thread-local variable called `LAST_ERROR`. The `get_error_message` function
+pulls the latest error message from `LAST_ERROR` and reports it to the
+user. All the mechanisms in `pact_matching_ffi/src/error` exist to handle
+this error collection and reporting.
+
+## Logging
+
+Logging is a crucial part of any application, and just because `pact_matching`
+is being called from another language is no reason to render logging info
+from it unavailable. For this reason, `pact_matching_ffi` provides a way for
+C callers to initialize a logger and direct logging output to a location of
+their choice.
+
+First, they call `logger_init`, which begins the logging setup by creating
+a log dispatcher with no sinks. Then they call `logger_attach_sink` to add
+sinks, to which any logs matching the filtering level will be sent. Finally,
+after all `logger_attach_sink` calls are done, they call `logger_apply` to
+apply the logger and complete setup.
+
+The remainder of the code in `pact_matching_ffi/src/log` is plumbing for this
+logging setup process.
+
+## The `ffi_fn` macro
+
+FFI functions have to be written with some boilerplate. First, they must be
+marked `#[no_mangle]` so their names are exported as-written for C to call.
+Second, they must be marked `pub extern` so they're exported and picked up
+by `cbindgen`, the tool we use for generating the C header which users of
+`pact_matching_ffi` will need.
+
+Then, within the functions themselves, we want to be sure that certain FFI
+information is logged, and that errors are captured and reported consistently
+and correctly.
+
+For this purpose, we use the `ffi_fn` macro, found in
+`pact_matching_ffi/src/util/ffi.rs`. This macro is inspired by a
+[macro of the same name][hyper_macro] from the `hyper` crate's C API
+(created for use of Hyper in `curl`). Our macro does considerably more work
+to capture errors and ensure complete logging, but the idea is the same.
+
+One thing to note is that most functions using the macro will be written
+slightly differently from normal Rust functions. If the function being
+wrapped is fallible, then a second block is needed after the function body
+block showing what to return in the case of any unexpected failure (usually
+either a null pointer of the appropriate `const`-ness and type, or a
+sentinel integer value signalling an error occurred).
+
+Here's an example of the macro in use.
+
+```rust
+ffi_fn! {
+    /// Get a mutable pointer to a newly-created default message on the heap.
+    fn message_new() -> *mut Message {
+        let message = Message::default();
+        ptr::raw_to(message)
+    } {
+        ptr::null_mut_to::<Message>()
+    }
+}
+```
+
+For destructors, or other functions which have to return value, no second
```

Comment:
Should this be "which have **_no_** return value"?

Reply:
Yup, good catch!